src/policy/compressor/forwarding.rs

```diff
@@ -0,0 +1,165 @@
+
+/// A finite-state machine which processes the positions of mark bits,
+/// and accumulates the size of live data that it has seen.
+///
+/// The Compressor caches the state of the transducer at the start of
+/// each block by serialising the state using [`Transducer::encode`];
+/// the state can then be deserialised using [`Transducer::decode`].
+#[derive(Debug)]
+struct Transducer {
+    live: usize,
+    last_bit_seen: Address,
+    in_object: bool,
+}
+impl Transducer {
+    pub fn new() -> Self {
+        Self {
+            live: 0,
+            last_bit_seen: Address::ZERO,
+            in_object: false,
+        }
+    }
+    pub fn step(&mut self, address: Address) {
+        if self.in_object {
+            self.live += address - self.last_bit_seen + MIN_OBJECT_SIZE;
+        }
+        self.in_object = !self.in_object;
+        self.last_bit_seen = address;
+    }
+
+    pub fn encode(&self, address: Address) -> usize {
+        if self.in_object {
+            // We count the space between the last mark bit and
+            // the current address as live when we stop in the
+            // middle of an object.
+            self.live + (address - self.last_bit_seen) + 1
+        } else {
+            self.live
+        }
+    }
+
+    pub fn decode(offset: usize, address: Address) -> Self {
+        Transducer {
+            live: offset & !1,
+            last_bit_seen: address,
+            in_object: (offset & 1) == 1,
+        }
+    }
+}
+
+pub struct ForwardingMetadata<VM: VMBinding> {
+    mark_bit_spec: SideMetadataSpec,
+    pub(crate) first_address: Address,
+    calculated: AtomicBool,
+    block_offsets: Vec<AtomicUsize>,
+    vm: PhantomData<VM>,
+}
+
+const BLOCK_SIZE: usize = 512;
+
+impl<VM: VMBinding> ForwardingMetadata<VM> {
+    pub fn new(
+        mark_bit_spec: SideMetadataSpec,
+        start: Address,
+        size: usize,
+    ) -> ForwardingMetadata<VM> {
+        let mut block_offsets = vec![];
+        let blocks = size / BLOCK_SIZE;
+        block_offsets.resize_with(blocks, || AtomicUsize::new(0));
+        ForwardingMetadata {
+            mark_bit_spec,
+            first_address: start,
+            calculated: AtomicBool::new(false),
+            block_offsets,
+            vm: PhantomData,
+        }
+    }
+
+    pub fn mark_end_of_object(&self, object: ObjectReference) {
+        use crate::util::metadata::side_metadata::{address_to_meta_address, meta_byte_lshift};
+        let end_of_object =
+            object.to_raw_address() + VM::VMObjectModel::get_current_size(object) - MIN_OBJECT_SIZE;
+        let a1 = address_to_meta_address(&self.mark_bit_spec, object.to_raw_address());
+        let s1 = meta_byte_lshift(&self.mark_bit_spec, object.to_raw_address());
+        let a2 = address_to_meta_address(&self.mark_bit_spec, end_of_object);
+        let s2 = meta_byte_lshift(&self.mark_bit_spec, end_of_object);
+        debug_assert!((a1, s1) < (a2, s2));
```

Comment:
Could you add a comment or print in the assertion to mention why this is required. It might not be clear to someone who's unfamiliar with the metadata layout in MMTk.